packages/ckeditor5-emoji/src/emojipicker.ts

```diff
@@ -0,0 +1,204 @@
+/**
+ * @license Copyright (c) 2003-2024, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module emoji/emojipicker
+ */
+
+import '../theme/emojipicker.css';
+import { icons, Plugin, type Editor } from 'ckeditor5/src/core.js';
+import { Picker } from 'emoji-picker-element';
+
+import {
+	clickOutsideHandler,
+	ContextualBalloon,
+	View,
+	ButtonView,
+	MenuBarMenuListItemButtonView
+} from 'ckeditor5/src/ui.js';
+
+type EmojiClickEvent = Event & {
+	detail: {
+		unicode: string;
+	};
+};
+
+/**
+ * The emoji picker plugin.
+ *
+ * Introduces the `'emoji'` dropdown.
+ */
+export default class EmojiPicker extends Plugin {
+	private _balloon!: ContextualBalloon;
+	declare private _emojiView: EmojiView;
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get requires() {
+		return [ ContextualBalloon ] as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'EmojiPicker' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static override get isOfficialPlugin(): true {
+		return true;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public init(): void {
+		this._balloon = this.editor.plugins.get( ContextualBalloon );
+		this._emojiView = this._createEmojiView();
+
+		this.editor.ui.componentFactory.add( 'emoji', () => this._createButton( ButtonView ) );
+		this.editor.ui.componentFactory.add( 'menuBar:emoji', () => this._createButton( MenuBarMenuListItemButtonView ) );
+	}
+
+	/**
+	 * Displays the balloon with the emoji picker.
+	 *
+	 * @param initialSearchValue Allows opening the emoji picker with given search query already typed in.
+	 */
+	public showUI( initialSearchValue?: string ): void {
+		this._balloon.add( {
+			view: this._emojiView,
+			position: this._getBalloonPositionData()
+		} );
+
+		this._emojiView.focus();
+
+		if ( initialSearchValue ) {
+			this._emojiView.updateSearchValue( initialSearchValue );
+		}
+	}
+
+	/**
+	 * Hides the balloon with the emoji picker.
+	 */
+	private _hideUI() {
+		this._balloon.remove( this._emojiView );
+
+		this.editor.editing.view.focus();
+	}
+
+	private _getBalloonPositionData() {
+		const view = this.editor.editing.view;
+		const viewDocument = view.document;
+
+		// Set a target position by converting view selection range to DOM.
+		const target = () => view.domConverter.viewRangeToDom( viewDocument.selection.getFirstRange()! );
+
+		return {
+			target
+		};
+	}
+
+	/**
+	 * Creates a (toolbar or menu bar) button for the emoji picker feature.
+	 */
+	private _createButton<T extends typeof ButtonView | typeof MenuBarMenuListItemButtonView>( ButtonClass: T ): InstanceType<T> {
```

Comment:
This could be an internal function that receives _this_ as a plugin instance. No need to expose this function as a method in class.